spatialos-sdk/src/worker/entity.rs

```diff
@@ -26,114 +25,56 @@ impl Entity {
-            },
-        );
+        self.pre_add_check(component_id)?;
+
+        self.components.insert(component_id, component);
 
         Ok(())
     }
 
-    pub(crate) unsafe fn add_serialized(
-        &mut self,
-        component_id: ComponentId,
-        mut component: Owned<SchemaComponentData>,
-    ) -> Result<(), String> {
-        let vtable = DATABASE.get_vtable(component_id).unwrap();
-        let deserialize_func = vtable
-            .worker_vtable
-            .component_data_deserialize
-            .unwrap_or_else(|| {
-                panic!(
-                    "No component_data_deserialize method defined for {}",
-                    component_id
-                )
-            });
-
-        // Create the **void that the C API requires. We need to then clean this up later.
-        // The value pointed to by handle_out_ptr is written to during the deserialize method.
-        let placeholder_ptr = Box::into_raw(Box::new(0)) as *mut ::std::os::raw::c_void;
-        let handle_out_ptr = Box::into_raw(Box::new(placeholder_ptr));
-
-        let deserialize_result = deserialize_func(
-            component_id,
-            ptr::null_mut(),
-            component.as_ptr_mut(),
-            handle_out_ptr,
-        );
-
-        match deserialize_result {
-            1 => {},
-            0 => return Err("Error deserializing serialized data. Is the SchemaComponentData malformed?".to_owned()),
-            _ => panic!("Unexpected return value from deserialize function. Expected true or false. Received other.")
-        };
-
-        let component_data = Worker_ComponentData {
-            reserved: ptr::null_mut(),
-            component_id,
-            schema_type: ptr::null_mut(),
-            user_handle: *handle_out_ptr,
-        };
-
-        // Reconstruct these objects so they can be de-alloc'ed. We have pulled the data required
-        // from them by de-referencing handle_out_ptr above.
-        Box::from_raw(placeholder_ptr);
-        Box::from_raw(handle_out_ptr);
-
-        self.add_raw(&component_data)
+    pub fn get<C: Component>(&self) -> Option<schema::Result<C>> {
+        self.components.get(&C::ID).map(|data| data.deserialize())
     }
 
-    pub fn get<C: Component>(&self) -> Option<&C> {
+    pub(crate) fn raw_component_data(&mut self) -> Vec<Worker_ComponentData> {
         self.components
-            .get(&C::ID)
-            .map(|data| unsafe { &*(data.user_handle as *const _) })
-    }
+            .iter_mut()
+            .map(|(&component_id, data)| Worker_ComponentData {
+                component_id,
 
-    pub(crate) fn raw_component_data(&self) -> RawEntity {
-        RawEntity::new(self.components.values())
+                // TODO: Why does this require a `*mut Schema_ComponentData`? Is there any actual
+                // chance that the underlying data will be mutated? Would it be safe for us to use
+                // `as_ptr` and cast it to a `*mut`?
```

Comment:
cc @dgavedissian ?

Reply:
Sorry, I'm not sure what this comment is referring to exactly?

Reply:
For a little bit better context:

```rust
Worker_ComponentData {
    component_id,

    // TODO: Why does this require a `*mut Schema_ComponentData`? Is there any actual
    // chance that the underlying data will be mutated? Would it be safe for us to use
    // `as_ptr` and cast it to a `*mut`?
    schema_type: data.as_ptr_mut(),

    ..Default::default()
}
```

`Worker_ComponentData::schema_type` is a `*mut Schema_ComponentData`, which means that `Entity::raw_component_data` needs to take `&mut self`. This seems odd, since in practice it doesn't seem like there's any reason why that data would actually be mutated. My assumption is that `Worker_ComponentData` is also used in cases where the value of `schema_type` *will* be mutated (e.g. vtable serialization), so the pointer has to be a `*mut`.

In practice, would it be safe to instead do `data.as_ptr() as *mut _`? For that to be the case, the SDK would have to not actually mutate the schema data. I thin the functions where we're passing this data are `Worker_Connection_SendCreateEntityRequest` and `Worker_SnapshotOutputStream_WriteEntity`.

Reply:
The more I think about it, the more confident that I feel that it should be safe to do `data.as_ptr() as *mut _`. I'm going to make that change now, but I'll hold off on merging until @dgavedissian can confirm.

Reply:
Note that in general, if you pass a `Schema_Whatever*` to a C API function, that function is actually going to take ownership over that data and free it when it's done with it. I haven't looked at the API in a while, but I believe that's the reason it's a mutable pointer.

See the following:
```
/**
 * Requests SpatialOS to create an entity. If components[i].schema_type is set, ownership is
 * transferred to the SDK, and components[i].schema_type is set to NULL. If
 * components[i].schema_type is NULL and components[i].user_handle is set, the entity data is
 * serialized immediately using the corresponding vtable serialize function.
 *
 * Returns Worker_RequestId -1 if the component is not registered and the default vtable is
 * not found.
 */
WORKER_API Worker_RequestId Worker_Connection_SendCreateEntityRequest(
    Worker_Connection* connection, uint32_t component_count, Worker_ComponentData* components,
    const Worker_EntityId* entity_id, const uint32_t* timeout_millis);
```

Reply:
Ah, excellent! I was looking for comments along that line explaining what would happening to `schema_type`, but I was looking in the wrong place (looking in the struct definition rather than the the function definition).

Reply:
Per your suggestion below, this comment is no longer necessary as we're now using `into_raw()` on the `SchemaComponentData` to get a `*mut Schema_ComponentData` :tada: